tools/sklearn/main_macros.xml

```diff
@@ -104,11 +101,80 @@ def get_X_y(params, file1, file2):
   return X, y
   </token>
 
+  <token name="@GET_SEARCH_PARAMS_FUNCTION@">
+def get_search_params(params_builder, in_pipeline=False):
+
+  search_params = {}
+
+  if not in_pipeline:
+    for ep in params_builder['estimator_param_set']:
+      if ep['estimator_param'].strip() == '':
+        continue
+      lst = ep['estimator_param'].split(":")
+      assert (len(lst) == 2), "Error, make sure there is one and only one colon in search parameter input."
+      literal = lst[1].strip().replace("__sq__","'")\
+          .replace('__dq__','"').replace("__oc__", "{").replace("__cc__", "}")\
+          .replace("__ob__", "[").replace("__cb__", "]")
+      try:
+        ev = ast.literal_eval(literal)
+      except ValueError:
+        valid = ['np.', 'numpy.', 'scipy.']
+        for v in valid:
+          if literal.startswith(v):
+            ev = eval(literal)
```

Comment:
~~Due to the `split(':')` above, I cannot currently generate any exploit but it does not give me confidence. If that was a `split(':', 1)`, then we would have an issue.~~ I was just being blind, here is an exploit:

```
literal = "numpy.array(__import__('subprocess').check_output('pwd'))"
```

So checking merely the starting characters isn't enough as you can immediately import things and start executing subprocesses. I do not have a lot of confidence in finding a safe way to validate calls to numpy/scipy. I wish I had better news/advice.

Reply:
Per discussion in gitter, blacklisting `'` or `"` isn't sufficient as you can construct strings from variables. e.g. `__name__[0] + __name__[3] + ...` and other variables.

Reply:
Thanks for your wonderful comment! My new thought is using regular expression to very the validity of the `literal`. Pattern could be like this: module(numpy/scipy).submodule.method(parameters). Could you please comment on this?

Reply:
The parameters part can take letter, number plus =.,()[]+'". We could block `_import_` as well.

Reply:
For GridSearchCV, I think we could potentially always use `list` datatype. `literal_eval` is enough.  For the randomized search, however, the advantage of random list generator over a fixed list seems  apparent. That's why we want the `eval` to be here. I am thinking we could further limit the generator within the `numpy.random` and `scipy.stats` two submodules. What do others think?